### PR TITLE
r/codedeploy_app - add arn attribute & tagging support & sweeper

### PR DESCRIPTION
**Pull request by DrFaust92**
Relates https://github.com/hashicorp/terraform-provider-aws/issues/8774

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSCodeDeployApp_'
--- PASS: TestAccAWSCodeDeployApp_disappears (29.33s)
--- PASS: TestAccAWSCodeDeployApp_computePlatform_Lambda (38.70s)
--- PASS: TestAccAWSCodeDeployApp_computePlatform_ECS (39.01s)
--- PASS: TestAccAWSCodeDeployApp_basic (44.33s)
--- PASS: TestAccAWSCodeDeployApp_computePlatform (61.79s)
--- PASS: TestAccAWSCodeDeployApp_name (64.22s)
--- PASS: TestAccAWSCodeDeployApp_tags (93.42s)
```

### CODE REVIEW

File: aws/resource_aws_codedeploy_app.go [hunk 1]
```diff
@@ -101,52 +110,74 @@ func resourceAwsCodeDeployAppCreate(d *schema.ResourceData, meta interface{}) er
...
 }
 
 func resourceAwsCodeDeployUpdate(d *schema.ResourceData, meta interface{}) error {
 	conn := meta.(*AWSClient).codedeployconn
 
-	o, n := d.GetChange("name")
+	if d.HasChange("tags") {
+		o, n := d.GetChange("tags")
 
-	_, err := conn.UpdateApplication(&codedeploy.UpdateApplicationInput{
```
Comment (anGie44) [inline aws/resource_aws_codedeploy_app.go:138]: it's interesting this was here given `name` is ForceNew...though testing this out, we could technically keep this update func, but need to remove the `ForceNew`, and update the ID during `Read` such that if `d.Id()` doesn't contain the value in `d.Get("name")`, it should be updated w/`d.SetId()`

File: aws/resource_aws_codedeploy_app.go [hunk 2]
```diff
@@ -101,52 +110,74 @@ func resourceAwsCodeDeployAppCreate(d *schema.ResourceData, meta interface{}) er
...
 		}
+
+		log.Printf("[ERROR] Error finding CodeDeploy application: %s", err)
+		return err
+	}
+
+	app := resp.Application
+	appName := aws.StringValue(app.ApplicationName)
+
+	appArn := arn.ARN{
```
Comment (anGie44) [inline aws/resource_aws_codedeploy_app.go:141]: i think we need to include this resource in the list found in `index.html` https://registry.terraform.io/providers/hashicorp/aws/latest/docs#skip_requesting_account_id

File: aws/resource_aws_codedeploy_app.go [hunk 3]
```diff
@@ -101,52 +110,74 @@ func resourceAwsCodeDeployAppCreate(d *schema.ResourceData, meta interface{}) er
...
+	d.SetId(fmt.Sprintf("%s:%s", aws.StringValue(resp.ApplicationId), application))
 
 	return resourceAwsCodeDeployAppRead(d, meta)
 }
 
 func resourceAwsCodeDeployAppRead(d *schema.ResourceData, meta interface{}) error {
 	conn := meta.(*AWSClient).codedeployconn
+	ignoreTagsConfig := meta.(*AWSClient).IgnoreTagsConfig
 
 	application := resourceAwsCodeDeployAppParseId(d.Id())
```
Comment (anGie44) [inline aws/resource_aws_codedeploy_app.go:122]: to support a change in ID suggested in the comment below:
```suggestion
        if !d.IsNewResource() && !strings.Contains(d.Id(), d.Get("name").(string)) {
		d.SetId(fmt.Sprintf("%s:%s", d.Get("application_id").(string), d.Get("name").(string)))
	}
	application := resourceAwsCodeDeployAppParseId(d.Id())
```

File: aws/resource_aws_codedeploy_app.go [hunk 4]
```diff
@@ -101,52 +110,74 @@ func resourceAwsCodeDeployAppCreate(d *schema.ResourceData, meta interface{}) er
...
-			return err
 		}
+
+		log.Printf("[ERROR] Error finding CodeDeploy application: %s", err)
+		return err
+	}
+
+	app := resp.Application
+	appName := aws.StringValue(app.ApplicationName)
+
```
Comment (anGie44) [inline aws/resource_aws_codedeploy_app.go:140]:
```suggestion

     if !strings.Contains(d.Id(), appName) {
         d.SetId(fmt.Sprintf("%s:%s", aws.StringValue(app.ApplicationId), appName))
     }
     
```

File: aws/resource_aws_codedeploy_app.go (hunk 2 above)
Comment (DrFaust92) [reply aws/resource_aws_codedeploy_app.go:141]: Forgot about that 😊

**Comment by DrFaust92** [comment]
Thanks @anGie44 addressed changes

```
--- PASS: TestAccAWSCodeDeployApp_disappears (27.04s)
--- PASS: TestAccAWSCodeDeployApp_computePlatform_Lambda (37.48s)
--- PASS: TestAccAWSCodeDeployApp_computePlatform_ECS (37.80s)
--- PASS: TestAccAWSCodeDeployApp_basic (44.69s)
--- PASS: TestAccAWSCodeDeployApp_computePlatform (59.50s)
--- PASS: TestAccAWSCodeDeployApp_name (63.87s)
--- PASS: TestAccAWSCodeDeployApp_tags (90.22s)
```

**Review by anGie44 — APPROVED** [review]
Thanks @DrFaust92 , looks great 👍 

Output of acceptance tests (commercial):
```
--- PASS: TestAccAWSCodeDeployApp_disappears (18.63s)
--- PASS: TestAccAWSCodeDeployApp_computePlatform_Lambda (23.43s)
--- PASS: TestAccAWSCodeDeployApp_computePlatform_ECS (25.15s)
--- PASS: TestAccAWSCodeDeployApp_basic (27.11s)
--- PASS: TestAccAWSCodeDeployApp_computePlatform (33.01s)
--- PASS: TestAccAWSCodeDeployApp_name (34.09s)
--- PASS: TestAccAWSCodeDeployApp_tags (44.21s)
```

Sweeper output:
```
go test ./aws -v -sweep=us-east-1,us-west-2 -sweep-run=aws_codedeploy_app -timeout 60m
2021/04/14 16:14:45 [DEBUG] Running Sweepers for region (us-east-1):
2021/04/14 16:14:45 [DEBUG] Running Sweeper (aws_codedeploy_app) in region (us-east-1)
2021/04/14 16:14:45 [INFO] AWS Auth provider used: "EnvProvider"
2021/04/14 16:14:45 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2021/04/14 16:14:45 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2021/04/14 16:14:46 Sweeper Tests ran successfully:
	- aws_codedeploy_app
2021/04/14 16:14:46 [DEBUG] Running Sweepers for region (us-west-2):
2021/04/14 16:14:46 [DEBUG] Running Sweeper (aws_codedeploy_app) in region (us-west-2)
2021/04/14 16:14:46 [INFO] AWS Auth provider used: "EnvProvider"
2021/04/14 16:14:46 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2021/04/14 16:14:46 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2021/04/14 16:14:47 Sweeper Tests ran successfully:
	- aws_codedeploy_app
```

**Pull request merged**